src/udev/udev-builtin-input_id.c

```diff
@@ -256,8 +258,11 @@ static bool test_pointers(sd_device *dev,
         if (is_tablet && has_pad_buttons)
                 is_tablet_pad = true;
 
+        if (has_pad_buttons && has_wheel)
+                is_tablet_pad = true;
```

Comment:
For historical reasons, `ID_INPUT_TABLET_PAD` is *in addition* to `ID_INPUT_TABLET`, so you'll need to set `is_tablet = true` here too. At which point the condition below doesn't need to be modified since it's covered by `!is_tablet` already.

Reply:
> 
> Note that libinput's `pad_init_buttons_from_kernel()` is only called where libwacom doesn't have an entry for that device (or libwacom is compiled out). It's definitely a fallback path, with a libwacom data file you can specify the button ordering. Mind you, that function should be fixed I reckon to take care of the gap between `BTN_9` and `BTN_MOUSE - 1`.
> 

Curiously I went and checked, and there are some XP-PEN devices defined in libwacom (the description never seems to make it up it's mind if it's only for Wacom devices.) I think it's good either way that it works with or without libwacom - but I'll be looking into adding support into libwacom for this tablet as well.

> I have [a list of all `libinput record` submitted to the libinput issue tracker](https://gitlab.freedesktop.org/libinput/libinput-recordings). Each recording has a back-link to the libinput issue it was filed for too as a comment at the top of the file.
> 
> With a bit of grepping we can get an idea, grep commands like this:
> 
> ```
> $ git grep -l BTN_STYLUS | sed -e 's|/.*\.recording$||' | uniq | sort > btn_stylus.devices
> $ git grep -L BTN_STYLUS | sed -e 's|/.*\.recording$||' | uniq | sort > not_btn_stylus.devices
> ```
> 

WOW that's incredibly useful, I wish I knew about that sooner. That also solves my worry that I'll be breaking other tablets without knowing.

> These are all pretty much the same, lots of buttons starting at `BTN_0` but no wheels. These are now I think these devices may be falling through the cracks now with no `ID_INPUT_*` tag assigned. This should be fixed by my comment regarding forcing `is_tablet = true` though.

I'll be changing that shortly :-)

Reply:
Fixed now, it's now marked as a tablet too.